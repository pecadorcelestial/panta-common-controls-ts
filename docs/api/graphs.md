# Button

### Properties

| Name    | Type   | Required | Comments |
| --------|--------|----------|----------|
| percent | number | true     | Sets the amount of the graphs that's going to be covered. |
| theme   | string | false    | Sets the theme (color scheme) for the button, these are the options currently available: ``red``, ``blue``, ``green``*, ``yellow``, ``gray``, ``orange`` and ``black``. |

**\*** _default_ value.
<br/>_Note: The **style** values for **width** and **float** can be overriden._

## Usage

```javascript
import React, { Component } from 'react';
import { CircularGraph } from 'panda-common-controls';

class BasicForm extends Component {
    render() {
        //Result.
        return(
            <div>
                <CircularGraph percent={85} theme='green'/>
                <CircularGraph percent={45} theme='orange'/>
                <CircularGraph percent={10} theme='red'/>
            <div>
        );
    }
}
```
