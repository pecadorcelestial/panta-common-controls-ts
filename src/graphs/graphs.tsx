//Modulos generales.
import React, { Component, HTMLProps } from 'react';
import styled, { css, keyframes, AnyStyledComponent } from 'styled-components';

type Theme = 'red' | 'blue' | 'green' | 'yellow' | 'gray' | 'orange' | 'black' | 'IENTC';
const backgroundColor = (theme: Theme): string => {
	switch(theme) {
		case 'red': return '#FF456A';
		case 'blue': return '#1476FB';
		case 'green': return '#4CB050';
		case 'yellow': return '#FFE200';
		case 'gray': return '#858585';
        case 'orange': return '#FF9700';
		case 'black': return '#242424';
        case 'IENTC': return '#FF0000';
	}
};

const Layout: AnyStyledComponent<HTMLProps<HTMLDivElement>, any> = styled.div`
    justify-content: space-around;
    width: 200px;
`;

const SVG: AnyStyledComponent<HTMLProps<SVGSVGElement>, any> = styled.svg`
    display: block;
    margin: 10px auto;
    max-height: 200px;
    max-width: 200px;
`;

const OutterPath: AnyStyledComponent<HTMLProps<SVGPathElement>, any> = styled.path`
    fill: none;
    stroke: #EEE;
    stroke-width: 2.8;
`;

const progress = (percent: number): any => keyframes`
    0% {
        stroke-dasharray: 0 100;
    }
    100% {
        stroke-dasharray: ${percent} 100;
    }
`;

const animationF = (props: IInnerPathProps) => css`
	${progress(props.percent)} 0.6s ease-out forwards;
`;

interface IInnerPathProps {
    theme: Theme;
    percent: number;
};
const InnerPath: AnyStyledComponent<IInnerPathProps & HTMLProps<SVGPathElement>, any> = styled.path`
    fill: none;
    stroke-linecap: round;
    stroke: ${(props: IInnerPathProps) => backgroundColor(props.theme)};
    stroke-width: 2.8;

    animation: ${animationF};
`;

const Text: AnyStyledComponent<IInnerPathProps & HTMLProps<HTMLTextAreaElement>, any> = styled.text`
    fill: #666;
    font-family: "Open Sans", sans-serif;
    font-size: 0.5em;
	font-stretch: normal;
	font-style: normal;	
	font-weight: bold;
    text-anchor: middle;
`;

interface ICircularGraphProps extends IInnerPathProps {};
export class CircularGraph extends Component<ICircularGraphProps & HTMLProps<HTMLDivElement>, {}> {
    render() {
        let { theme, percent, ...rest } = this.props;
        return(
            <Layout {...rest}>
                <SVG viewBox="0 0 36 36">
                    <OutterPath
                        d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
                    />
                    <InnerPath
                        percent={percent}
                        theme={theme}
                        d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
                    />
                    <Text x="18" y="20.35">{`${percent}%`}</Text>
                </SVG>
            </Layout>
        );
    }
}