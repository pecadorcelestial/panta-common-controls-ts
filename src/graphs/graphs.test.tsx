
//RRRR  EEEEE  AAA   CCCC TTTTT
//R   R E     A   A C       T
//RRRR  EEE   AAAAA C       T
//R   R E     A   A C       T
//R   R EEEEE A   A  CCCC   T

import React from 'react';
import renderer from 'react-test-renderer';

//EEEEE N   N ZZZZZ Y   Y M   M EEEEE
//E     NN  N    Z   Y Y  MM MM E
//EEE   N N N   Z     Y   M M M EEE
//E     N  NN  Z      Y   M   M E
//EEEEE N   N ZZZZZ  YYY  M   M EEEEE

import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });

//JJJJJ EEEEE  SSSS TTTTT
//    J E     S       T
//    J EEE    SSS    T
//J   J E         S   T
// JJJ  EEEEE SSSS    T

//NOTA: Éste módulo ayuda a que los errores en componentes estilizados se vean de manera correcta:
//https://medium.com/styled-components/effective-testing-for-styled-components-67982eb7d42b
import 'jest-enzyme';
import 'jest-styled-components';

// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT EEEEE  SSSS
//C     O   O MM MM P   P O   O NN  N E     NN  N   T   E     S
//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   EEE    SSS
//C     O   O M   M P     O   O N  NN E     N  NN   T   E         S
// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   EEEEE SSSS

import { CircularGraph } from './graphs';

//Snapshots (Temas: red, secondaryRed, flatRed, blue, flatBlue, green, flatGreen, yellow, flatYellow, gray, flatGray).
describe('[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black.', () => {
	//Tema: red.
	it('Debe pintar el componente con el tema "red".', () => {
        //Snapshot [red].
        let component = renderer.create(<CircularGraph theme='red' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Tema: blue.
	it('Debe pintar el componente con el tema "blue".', () => {
        //Snapshot [blue].
        let component = renderer.create(<CircularGraph theme='blue' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
	});
    //Tema: green.
	it('Debe pintar el componente con el tema "green".', () => {
        //Snapshot [green].
        let component = renderer.create(<CircularGraph theme='green' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Tema: yellow.
	it('Debe pintar el componente con el tema "yellow".', () => {
        //Snapshot [yellow].
        let component = renderer.create(<CircularGraph theme='yellow' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Tema: gray.
	it('Debe pintar el componente con el tema "gray".', () => {
        //Snapshot [gray].
        let component = renderer.create(<CircularGraph theme='gray' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Tema: orange.
	it('Debe pintar el componente con el tema "orange".', () => {
        //Snapshot [orange].
        let component = renderer.create(<CircularGraph theme='orange' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Tema: black.
	it('Debe pintar el componente con el tema "black".', () => {
        //Snapshot [black].
        let component = renderer.create(<CircularGraph theme='black' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Tema: ientc.
	it('Debe pintar el componente con el tema "ientc".', () => {
        //Snapshot [ientc].
        let component = renderer.create(<CircularGraph theme='IENTC' percent={100}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
});