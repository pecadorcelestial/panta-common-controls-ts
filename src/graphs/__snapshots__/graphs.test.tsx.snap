// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "black". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #242424;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "blue". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #1476FB;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "gray". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #858585;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "green". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #4CB050;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "ientc". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #FF0000;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "orange". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #FF9700;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "red". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #FF456A;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CircularGraph] - Gráfica de círculo con los siguientes temas: red, blue, green, yellow, gray, orange y black. Debe pintar el componente con el tema "yellow". 1`] = `
.c0 {
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 200px;
}

.c1 {
  display: block;
  margin: 10px auto;
  max-height: 200px;
  max-width: 200px;
}

.c2 {
  fill: none;
  stroke: #EEE;
  stroke-width: 2.8;
}

.c3 {
  fill: none;
  stroke-linecap: round;
  stroke: #FFE200;
  stroke-width: 2.8;
  -webkit-animation: czDoju 0.6s ease-out forwards;
  animation: czDoju 0.6s ease-out forwards;
}

.c4 {
  fill: #666;
  font-family: "Open Sans",sans-serif;
  font-size: 0.5em;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  text-anchor: middle;
}

<div
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 36 36"
  >
    <path
      className="c2"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <path
      className="c3"
      d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
    />
    <text
      className="c4"
      x="18"
      y="20.35"
    >
      100%
    </text>
  </svg>
</div>
`;
